state-transition/core/state_processor_withdrawals.go

```
@@ -136,3 +155,190 @@ func (sp *StateProcessor) processWithdrawals(
 
 	return nil
 }
+
+// processWithdrawalRequest is the equivalent of process_withdrawal_request as defined in the spec.
+// It should only be called after the electra hard fork.
+// For invalid withdrawal requests, we return nil, and only return error for system errors.
+func (sp *StateProcessor) processWithdrawalRequest(st *state.StateDB, withdrawalRequest *ctypes.WithdrawalRequest) error {
+	amount := withdrawalRequest.Amount
+	// If the amount is 0, it's a full exit.
+	isFullExitRequest := amount == constants.FullExitRequestAmount
+	pendingPartialWithdrawals, err := st.GetPendingPartialWithdrawals()
+	if err != nil {
+		return err
+	}
+	// If partial withdrawal queue is full, only full exits are processed
+	if len(pendingPartialWithdrawals) == constants.PendingPartialWithdrawalsLimit && !isFullExitRequest {
+		sp.logger.Warn(
+			"skipping processing of withdrawal request as partial withdrawal queue is full",
+			withdrawalFields(withdrawalRequest, nil)...,
+		)
+		return nil
+	}
```

question: this means we skip for now, but we will get to it in the next block right? since the next block will process another 8 partial withdrawals

---

no it will dropped

---

this is the edge case the concerns me the most with regards to UX as it silently dropping without any mistake from a withdrawer

---

ah yeah, its never stored in the state. We should add a metric as we do for [non-withdrawable deposits](https://github.com/berachain/beacon-kit/blob/main/state-transition/core/state_processor_staking.go#L145) and [deposit stake lost](https://github.com/berachain/beacon-kit/blob/main/state-transition/core/state_processor_staking.go#L167)

---

Yea this PR doesn't have any metrics atm. Will save it for follow up PRs as we go - a dashboard for the logs would suffice for now

---

Should be done before cutting an rc so the metric can be tracked as it would be on mainnet

---

Should probably also have a warning/alerting if we are approaching close to this limit, to alert externally not to create withdrawals

---

Agreed - our alerting could use a lot of improvement.

---

https://github.com/berachain/beacon-kit/issues/2710